decisions/2020/06/25-continuous-app-delivery-discussion.md

```diff
@@ -0,0 +1,81 @@
+# 2020-06-25, Continuous App Delivery
+
+## Attending
+
+Austin, Erik, Lars, Stian, Zubair, Birk
+
+## Context
+
+The meeting was scheduled to identify and propose solutions for features that we need to support continuous app delivery for core applications.  
+
+Link to related issue and agenda: https://github.com/dhis2/notes/issues/114
+
+
+## Support multiple versions of the same app
+
+- This was top priority before the meeting, as we wanted to have admins test apps on the instance before installing a new versions.
+- People are accustomed to have a staging server, so the need for this is not as urgent.
+
+**Conclusion:**
+
+- This has been postponed for now as there are more urgent issues.
+- A nice to have down the line.
+
+
+## Support for overriding core apps by custom apps
+
+- To be able to update core apps we need them to install them as a "custom app".
+- The proposed solution was to have the server check for a custom app of the same name, and use that if it exists.
+- We need to support "old" urls (eg. "dhis-web-dashboard") for backwards compatability, these should redirect to a possibly new url (eg. "/apps")
+- The issue of collisions of app names were brought up, as users could unintentionally install apps by the same name of the core app, thus overriding it.
+- Installing apps should be a locked down permission, and admins should be aware of conflicts.
+- We could potentially notify the user the first time a "custom core app" is installed.
+
+**Conclusion:**
+
+- Going forward with implementation of overrides of core apps.
+- Jira issue for this: https://jira.dhis2.org/browse/DHIS2-4394
+
+## Re-introduce the proxy through core to App Hub
+
+- The proxy was removed mostly due to the ease of introducing new APIs between the App Hub and management app. 
+- Concerns were raised of the removal of the proxy to App Hub.
+    - Air gaps, as clients may not have internet access or have poor internet connections.
```

Comment:
Does this mean that we have a requirement to support situations where the DHIS2 instance **has** internet access, and the clients using the DHIS2 instance **does not**?

Reply:
Yes, as I understood it there are implementations where only the server has internet access. We also have to be careful here, as this would be the first time (as far as I know) a core app makes external requests. 

My train of thought was also that installing apps is an admin-only use-case, and they probably have an internet connection. But if this assumption is false, we should reimplement the proxy.

Reply:
@varl that's correct.  The specific use-case I have heard is federated MSF DHIS2 instances in rural clinics which will periodically activate a satellite internet connection on the DHIS2 server to do data upload and metadata synchronization.  It is certainly possible that in these rural locations there is no mobile coverage, so clients also do not have any reliable internet connection, or their connection is slow and expensive.  In this case the ideal architecture would be to support application bundle download **from the server directly** after being initiated by an admin-only API call from a local client.

Reply:
Well, that is currently how it works when installing apps. The client -> App hub direct connection is used to get list of apps and channels. The App Management app initiates an install by a post with the ID of the app, eg. `https://play.dhis2.org/dev/api/appHub/a9adfa34-2bf9-4179-b692-43712d5632e7`. The client does **not** download the bundle and then upload it to the instance to install it.

Reply:
That makes sense, it would be a shame to lock those instances out of core app updates.

> In this case the ideal architecture would be to support application bundle download from the server directly after being initiated by an admin-only API call from a local client.

Yeah, that's what we do to install an app now.

So if I am understanding it correctly, the specific task in JIRA-9093 is no more than:

- Update to version 2 App Hub API end-points in the Core API: https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/apphub/DefaultAppHubService.java#L80-L88

- Change App Management App to get list of apps from Core API instead of App Hub API: https://github.com/dhis2/app-management-app/blob/master/src/actions.js#L89-L127

Reply:
> If I am now up-to-speed and grasp the discussion from the meeting that doesn't sound too bad, though the drawback of lagging 6 months on App Hub features being implemented in the Core API (and not backported to older DHIS2 versions since they are new features) are real.

I discussed using a wildcard App Hub proxy in the core with @larshelge today and he agreed it shouldn't be an issue to do that.  We should run it by @netroms as well to confirm, but it sounds like having a pass-through proxy (while actually installing the apps with a well-known API from within the core) should be fine 👍

Reply:
Awesome that we can evaluate that as an option, thanks for following up directly. 👍

Reply:
Great, @amcgee ! I will update the minutes and Jira-issue with that!